src/target/riscv/riscv-013.c

```diff
@@ -1145,15 +1171,28 @@ static int register_write_abstract(struct target *target, enum gdb_regno number,
 			!info->abstract_write_csr_supported)
 		return ERROR_FAIL;
 
-	uint32_t command = access_register_command(target, number, size,
+	const unsigned int size_bits = register_size(target, number);
+	const uint32_t command = access_register_command(target, number, size_bits,
 			AC_ACCESS_REGISTER_TRANSFER |
 			AC_ACCESS_REGISTER_WRITE);
+	LOG_DEBUG_REG(target, AC_ACCESS_REGISTER, command);
+	assert(size_bits % 32 == 0);
+	const unsigned int size_in_words = size_bits / 32;
+	/** The batch consists of:
+	 * * A write of "data*" ("size_in_words" scans).
+	 * * A write of "command" (1 scan).
+	 * * A read of "abstractcs" (1 scan).
+	 * * An implicit "nop".
+	 */
+	const unsigned int batch_size = size_in_words + 2;
+	struct riscv_batch * const batch = riscv_batch_alloc(target, batch_size);
 
-	if (write_abstract_arg(target, 0, value, size) != ERROR_OK)
-		return ERROR_FAIL;
-
+	abstract_data_write_fill_batch(batch, value, /*index*/ 0, size_bits);
 	uint32_t cmderr;
-	int result = execute_abstract_command(target, command, &cmderr);
+	const int result = execute_abstract_command_in_batch(target, batch, command,
+			&cmderr);
+	riscv_batch_free(batch);
```

Comment:
I am afraid the division of the responsibilities between the functions, as it is proposed now, makes the code harder to understand:

- The caller must allocate the batch - it must know the proper size, but it is not the caller who fills the scan items. 
- Abstract_data_write_fill_batch() adds the scans to write dataN registers. This is ok - one function does one thing.
- Execute_abstract_command_in_batch() does too many things - it adds further scans, then executes the batch, and finally it checks the abstract command result.


For these reasons, please consider different approach: Create multiple functions that will handle the execution of the whole abstract command batches. It would be:

- `execute_abstract_cmd_as_batch_w0(...)` - allocate the batch of needed size, fill the batch with write to `arg0`, write to command, then execute the batch, then check the result and finally free the batch
- `execute_abstract_cmd_as_batch_r0(...)` - same as above, but instead of writing to `arg0`, read `arg0` after the command is executed
- `execute_abstract_cmd_as_batch_w1r0(...)` - same, but write `arg1` before the command and read `arg0` after (needed for abstract memory read)
- `execute_abstract_cmd_as_batch_w1w0(...)` - same, but write `arg1` and `arg0` before the command (needed for abstract memory write)
- `execute_abstract_cmd_as_batch(...)` - only execute the abstract command and don't care about reading or writing the abstract arguments (dataN registers)

(The naming convention can of course be different - I just tried to present the idea.)

This way the functions would handle all the processing, including the management of the batches. And the code of the caller would be cleaner, shorter and easier to read - more focused on the actual algorithm, not at details of abstract commands or batches.

Reply:
You are right, the function was poorly designed.
Please, take a look at current approach.

I didn't go forward with your suggestion since there is an issue.

Consider an example: `execute_abstract_cmd_as_batch_w1[r/w]0()` used to access memory via abstract command in case address auto increment is not supported.
Here it is actually neaded to merge as much consequent abstract command executions into a single batch, since it's a memory read/write -- a long operation.
With the approach you suggest it is not possible.

Reply:
Good point about the abstract memory access and the issue with my proposal - I agree.

The current approach looks good to me. Thank you.

I have only proposed different names for some of the new functions. See one of my other comments.